File: src/pci/virtio/blk/blk.c
#include "blk.h"
#include "mm/mmu/mmu.h"
#include "mmio/mmio.h"
#include "uart/uart.h"
#include "utils/utils.h"

struct virtio_blk blk_dev;

/* Page-aligned backing memory for the virtqueue */
static struct virtq_desc blk_desc[VIRTQ_MAX_SIZE]
    __attribute__((aligned(4096)));
static struct virtq_avail blk_avail __attribute__((aligned(4096)));
static struct virtq_used blk_used __attribute__((aligned(4096)));

void pci_virtio_blk_init(void) {
  uart_println("[BLK] Initializing Device");

  /* Step 0: Find device */
  if (!pci_find_device(VIRTIO_BLK_VENDOR_ID, VIRTIO_BLK_DEVICE_ID,
                       &blk_dev.pci)) {
    uart_errorln("[BLK] Device not found");
    return;
  }

  uart_println("[BLK] Device found");

  // Header Type register[6:0] = header layout type
  // https://wiki.osdev.org/PCI#Header_Type_Register
  if ((pci_get_header_type(&blk_dev.pci) & 0x7F) != PCI_ENDPOINT_DEV_TYPE) {
    uart_errorln("[BLK]: Unexpected header type");
    return;
  }

  pci_assign_bars(&blk_dev.pci);
  pci_enable_device(&blk_dev.pci);
  virtio_parse_capabilities(&blk_dev.pci, &blk_dev.pci_caps);

  /* VirtIO Device Init Sequence
   * All register accesses go through the MMIO layer. */
  uintptr_t base = blk_dev.pci_caps.common_cfg;

  /* Step 1: Reset Device */
  uart_println("[BLK][VIRTIO-INIT][1] Reset Device");
  mmio_write8(base + VIRTIO_COMMON_STATUS, VIRTIO_STATUS_RESET);
  dsb_sy();

  /* Wait for RESET to complete */
  while (mmio_read8(base + VIRTIO_COMMON_STATUS) != VIRTIO_STATUS_RESET) {
  }
  uart_println("[BLK][VIRTIO-INIT][1] Reset Device Complete");

  /* Step 2: ACK */
  uart_println("[BLK][VIRTIO-INIT][2] Ack");
  uint8_t status = mmio_read8(base + VIRTIO_COMMON_STATUS);
  mmio_write8(base + VIRTIO_COMMON_STATUS, status | VIRTIO_STATUS_ACKNOWLEDGE);
  dsb_sy();

  /* Step 3: Set Driver status */
  uart_println("[BLK][VIRTIO-INIT][3] Driver Status");
  status = mmio_read8(base + VIRTIO_COMMON_STATUS);
  mmio_write8(base + VIRTIO_COMMON_STATUS, status | VIRTIO_STATUS_DRIVER);
  dsb_sy();

  /* Step 4: Feature Negotiation */
  uart_println("[BLK][VIRTIO-INIT][4] Negotiate Features");

  /* Read device features - low 32 bits */
  mmio_write32(base + VIRTIO_COMMON_DFSELECT, 0);
  dsb_sy();
  uint32_t feat_lo = mmio_read32(base + VIRTIO_COMMON_DF);
  uart_printf(" Device features[0]: %x\n", feat_lo);

  /* Read device features - high 32 bits */
  mmio_write32(base + VIRTIO_COMMON_DFSELECT, 1);
  dsb_sy();
  uint32_t feat_hi = mmio_read32(base + VIRTIO_COMMON_DF);
  uart_printf(" Device features[1]: %x\n", feat_hi);

  /* accept only VIRTIO_F_VERSION_1 (bit 32, in feat_hi bit 0) */
  uint32_t guest_lo = 0;
  uint32_t guest_hi = feat_hi & 0x01;

  mmio_write32(base + VIRTIO_COMMON_GFSELECT, 0);
  dsb_sy();
  mmio_write32(base + VIRTIO_COMMON_GF, guest_lo);
  dsb_sy();
  mmio_write32(base + VIRTIO_COMMON_GFSELECT, 1);
  dsb_sy();
  mmio_write32(base + VIRTIO_COMMON_GF, guest_hi);
  dsb_sy();

  uart_printf(" Accepted Features: lo=%x hi=%x\n", guest_lo, guest_hi);

  /* Step 5: FEATURES_OK */
  status = mmio_read8(base + VIRTIO_COMMON_STATUS);
  mmio_write8(base + VIRTIO_COMMON_STATUS, status | VIRTIO_STATUS_FEATURES_OK);
  dsb_sy();

  /* Step 6: Re-read and verify FEATURES_OK stuck */
  status = mmio_read8(base + VIRTIO_COMMON_STATUS);
  if (!(status & VIRTIO_STATUS_FEATURES_OK)) {
    uart_errorln("[BLK] FEATURES_OK failed");
    return;
  }
  uart_printf("[BLK] Status: %x\n", (uint32_t)status);
  uart_println("[BLK] FEATURES_OK !");

  /* Step 6: Setup virtqueue 0 */
  blk_dev.vq.desc = blk_desc;
  blk_dev.vq.avail = &blk_avail;
  blk_dev.vq.used = &blk_used;

  if (virtqueue_setup(base, 0, &blk_dev.vq, &blk_dev.pci_caps) != ESUCCESS) {
    uart_errorln("[BLK] Virtqueue setup failed");
    return;
  }

  /* Step 7: DRIVER_OK */
  status = mmio_read8(base + VIRTIO_COMMON_STATUS);
  mmio_write8(base + VIRTIO_COMMON_STATUS, status | VIRTIO_STATUS_DRIVER_OK);
  dsb_sy();
  uart_println("[BLK] DRIVER_OK set");

  /* Read capacity from device config (u64, sectors) */
  uintptr_t dcfg = blk_dev.pci_caps.device_cfg;
  uint32_t cap_lo = mmio_read32(dcfg + VIRTIO_BLK_CFG_CAPACITY);
  uint32_t cap_hi = mmio_read32(dcfg + VIRTIO_BLK_CFG_CAPACITY + 4);
  blk_dev.capacity_sectors = ((uint64_t)cap_hi << 32) | cap_lo;
  uart_printf("[BLK] Capacity: %d sectors (%d MiB)\n",
              (uint64_t)blk_dev.capacity_sectors,
              (uint64_t)(blk_dev.capacity_sectors / 2048));

  return;
}

int blk_read(uint64_t sector, void *buf) {
  static struct virtio_blk_req hdr __attribute__((aligned(16)));
  static volatile uint8_t status __attribute__((aligned(16)));

  hdr.type = VIRTIO_BLK_T_IN;
  hdr.reserved = 0;
  hdr.sector = sector;
  status = 0xFF;

  struct virtq_seg segs[3] = {
      {VIRT_TO_PHYS((uint64_t)(uintptr_t)&hdr), sizeof(hdr),
       VIRTQ_DESC_F_NONE},
      {VIRT_TO_PHYS((uint64_t)(uintptr_t)buf), VIRTIO_BLK_SECTOR_SIZE,
       VIRTQ_DESC_F_WRITE},
      {VIRT_TO_PHYS((uint64_t)(uintptr_t)&status), 1, VIRTQ_DESC_F_WRITE},
  };

  virtqueue_submit_chain(&blk_dev.vq, segs, 3);
  virtqueue_notify(&blk_dev.vq);
  virtqueue_poll(&blk_dev.vq);

  if (status != VIRTIO_BLK_S_OK) {
    uart_printf("[BLK] read sector %d failed: status=%x\n", sector,
                (uint32_t)status);
    return EERROR;
  }
  return ESUCCESS;
}

int blk_write(uint64_t sector, const void *buf) {
  static struct virtio_blk_req hdr __attribute__((aligned(16)));
  static volatile uint8_t status __attribute__((aligned(16)));

  hdr.type = VIRTIO_BLK_T_OUT;
  hdr.reserved = 0;
  hdr.sector = sector;
  status = 0xFF;

  struct virtq_seg segs[3] = {
      {VIRT_TO_PHYS((uint64_t)(uintptr_t)&hdr), sizeof(hdr),
       VIRTQ_DESC_F_NONE},
      {VIRT_TO_PHYS((uint64_t)(uintptr_t)buf), VIRTIO_BLK_SECTOR_SIZE,
       VIRTQ_DESC_F_NONE},
      {VIRT_TO_PHYS((uint64_t)(uintptr_t)&status), 1, VIRTQ_DESC_F_WRITE},
  };

  virtqueue_submit_chain(&blk_dev.vq, segs, 3);
  virtqueue_notify(&blk_dev.vq);
  virtqueue_poll(&blk_dev.vq);

  if (status != VIRTIO_BLK_S_OK) {
    uart_printf("[BLK] write sector %d failed: status=%x\n", sector,
                (uint32_t)status);
    return EERROR;
  }
  return ESUCCESS;
}

File: src/virtio/blk.rs
//! VirtIO block driver — synchronous 512-byte sector read/write.
//!
//! Port of `src/pci/virtio/blk/blk.c`.

use super::virtqueue::{self, Virtqueue, VirtqAvail, VirtqDesc, VirtqSeg, VirtqUsed, VirtqUsedElem, VIRTQ_DESC_F_WRITE, VIRTQ_MAX_SIZE};
use super::*;
use crate::kprintln;
use crate::mm::mmu::virt_to_phys;
use crate::mmio;
use crate::pci;
use crate::sync::{Racy, SpinLock};
use crate::uart;

pub const VIRTIO_BLK_VENDOR_ID: u16 = 0x1AF4;
pub const VIRTIO_BLK_DEVICE_ID: u16 = 0x1042;
pub const VIRTIO_BLK_SECTOR_SIZE: usize = 512;

const VIRTIO_BLK_CFG_CAPACITY: usize = 0x00;
const VIRTIO_BLK_T_IN: u32 = 0;
const VIRTIO_BLK_T_OUT: u32 = 1;
const VIRTIO_BLK_S_OK: u8 = 0;

#[repr(C, align(16))]
struct BlkReq {
    type_: u32,
    reserved: u32,
    sector: u64,
}

#[repr(C, align(4096))]
struct BlkBacking {
    desc: [VirtqDesc; VIRTQ_MAX_SIZE],
    avail: VirtqAvail,
    used: VirtqUsed,
    hdr: BlkReq,
    status: u8,
    _pad: [u8; 15],
}

static BACKING: Racy<BlkBacking> = Racy::new(BlkBacking {
    desc: [VirtqDesc { addr: 0, len: 0, flags: 0, next: 0 }; VIRTQ_MAX_SIZE],
    avail: VirtqAvail { flags: 0, idx: 0, ring: [0; VIRTQ_MAX_SIZE] },
    used: VirtqUsed { flags: 0, idx: 0, ring: [VirtqUsedElem { id: 0, len: 0 }; VIRTQ_MAX_SIZE] },
    hdr: BlkReq { type_: 0, reserved: 0, sector: 0 },
    status: 0,
    _pad: [0; 15],
});

struct BlkDev {
    caps: VirtioPciCaps,
    vq: Virtqueue,
    capacity_sectors: u64,
    ready: bool,
}

static BLK: Racy<BlkDev> = Racy::new(BlkDev {
    caps: VirtioPciCaps { common_cfg: 0, notify_base: 0, isr_cfg: 0, device_cfg: 0, notify_off_multiplier: 0 },
    vq: Virtqueue::empty(),
    capacity_sectors: 0,
    ready: false,
});

pub fn init() {
    uart::println("[BLK] Initializing Device");
    let mut dev = match pci::find_device(VIRTIO_BLK_VENDOR_ID, VIRTIO_BLK_DEVICE_ID) {
        Some(d) => d,
        None => {
            uart::errorln("[BLK] Device not found");
            return;
        }
    };
    if pci::header_type(&dev) & 0x7F != pci::PCI_ENDPOINT_DEV_TYPE {
        uart::errorln("[BLK] Unexpected header type");
        return;
    }
    pci::assign_bars(&mut dev);
    pci::enable_device(&dev);

    let blk = unsafe { BLK.get() };
    super::parse_capabilities(&dev, &mut blk.caps);
    let base = blk.caps.common_cfg;

    mmio::write8(base + VIRTIO_COMMON_STATUS, VIRTIO_STATUS_RESET);
    dsb_sy();
    while mmio::read8(base + VIRTIO_COMMON_STATUS) != VIRTIO_STATUS_RESET {}
    let mut status = mmio::read8(base + VIRTIO_COMMON_STATUS);
    mmio::write8(base + VIRTIO_COMMON_STATUS, status | VIRTIO_STATUS_ACKNOWLEDGE);
    dsb_sy();
    status = mmio::read8(base + VIRTIO_COMMON_STATUS);
    mmio::write8(base + VIRTIO_COMMON_STATUS, status | VIRTIO_STATUS_DRIVER);
    dsb_sy();

    mmio::write32(base + VIRTIO_COMMON_DFSELECT, 1);
    dsb_sy();
    let feat_hi = mmio::read32(base + VIRTIO_COMMON_DF);
    let guest_hi = feat_hi & 0x01;
    mmio::write32(base + VIRTIO_COMMON_GFSELECT, 0);
    dsb_sy();
    mmio::write32(base + VIRTIO_COMMON_GF, 0);
    dsb_sy();
    mmio::write32(base + VIRTIO_COMMON_GFSELECT, 1);
    dsb_sy();
    mmio::write32(base + VIRTIO_COMMON_GF, guest_hi);
    dsb_sy();

    status = mmio::read8(base + VIRTIO_COMMON_STATUS);
    mmio::write8(base + VIRTIO_COMMON_STATUS, status | VIRTIO_STATUS_FEATURES_OK);
    dsb_sy();
    status = mmio::read8(base + VIRTIO_COMMON_STATUS);
    if status & VIRTIO_STATUS_FEATURES_OK == 0 {
        uart::errorln("[BLK] FEATURES_OK failed");
        return;
    }

    let backing = unsafe { BACKING.get() };
    blk.vq.desc = backing.desc.as_mut_ptr();
    blk.vq.avail = &mut backing.avail as *mut VirtqAvail;
    blk.vq.used = &mut backing.used as *mut VirtqUsed;
    if !virtqueue::setup(base, 0, &mut blk.vq, &blk.caps) {
        uart::errorln("[BLK] Virtqueue setup failed");
        return;
    }

    status = mmio::read8(base + VIRTIO_COMMON_STATUS);
    mmio::write8(base + VIRTIO_COMMON_STATUS, status | VIRTIO_STATUS_DRIVER_OK);
    dsb_sy();

    let dcfg = blk.caps.device_cfg;
    let cap_lo = mmio::read32(dcfg + VIRTIO_BLK_CFG_CAPACITY) as u64;
    let cap_hi = mmio::read32(dcfg + VIRTIO_BLK_CFG_CAPACITY + 4) as u64;
    blk.capacity_sectors = (cap_hi << 32) | cap_lo;
    blk.ready = true;
    kprintln!(
        "[BLK] DRIVER_OK; capacity {} sectors ({} MiB)",
        blk.capacity_sectors,
        blk.capacity_sectors / 2048
    );
}

pub fn capacity_sectors() -> u64 {
    unsafe { BLK.get() }.capacity_sectors
}

pub fn is_ready() -> bool {
    unsafe { BLK.get() }.ready
}

/// Serializes block requests: there is a single shared request header and
/// virtqueue, so concurrent callers (e.g. boot ELF loads racing shell FS ops)
/// would corrupt it. Busy-poll completion means a plain spinlock is safe.
static BLK_LOCK: SpinLock<()> = SpinLock::new(());

fn rw(sector: u64, buf_pa: u64, write: bool) -> bool {
    let _guard = BLK_LOCK.lock();
    let blk = unsafe { BLK.get() };
    if !blk.ready {
        return false;
    }
    let backing = unsafe { BACKING.get() };
    backing.hdr.type_ = if write { VIRTIO_BLK_T_OUT } else { VIRTIO_BLK_T_IN };
    backing.hdr.reserved = 0;
    backing.hdr.sector = sector;
    backing.status = 0xFF;

    let hdr_pa = virt_to_phys(&backing.hdr as *const BlkReq as u64);
    let status_pa = virt_to_phys(&backing.status as *const u8 as u64);
    let data_flags = if write { 0 } else { VIRTQ_DESC_F_WRITE };
    let segs = [
        VirtqSeg { pa: hdr_pa, len: core::mem::size_of::<BlkReq>() as u32, flags: 0 },
        VirtqSeg { pa: buf_pa, len: VIRTIO_BLK_SECTOR_SIZE as u32, flags: data_flags },
        VirtqSeg { pa: status_pa, len: 1, flags: VIRTQ_DESC_F_WRITE },
    ];
    virtqueue::submit_chain(&mut blk.vq, &segs);
    virtqueue::notify(&blk.vq);
    virtqueue::poll(&mut blk.vq);

    if backing.status != VIRTIO_BLK_S_OK {
        kprintln!("[BLK] sector {} io failed status={:#x}", sector, backing.status);
        return false;
    }
    true
}

/// Read one 512-byte sector into `buf` (must be >= 512 bytes, kernel-mapped).
pub fn read(sector: u64, buf: &mut [u8]) -> bool {
    let pa = virt_to_phys(buf.as_ptr() as u64);
    let ok = rw(sector, pa, false);
    ok
}

/// Write one 512-byte sector from `buf`.
pub fn write(sector: u64, buf: &[u8]) -> bool {
    let pa = virt_to_phys(buf.as_ptr() as u64);
    rw(sector, pa, true)
}
